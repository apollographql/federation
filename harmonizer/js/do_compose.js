/** @typedef {{sdl: string, name: string, url?: string;}} ServiceDefinition */

/**
 * This `composition` is defined as a global by the runtime we define in Rust.
 * We declare this as a `var` here only to allow the TSDoc type annotation to be
 * applied to it. Running `var` multiple times has no effect.
 * @type {{
 *   composeServices: import('../../composition-js').composeServices,
 *   parseGraphqlDocument: import('graphql').parse
 * }} */
var composition;

/**
 * @type {ServiceDefinition[]}
 */
var serviceList = serviceList;

if (!serviceList || !Array.isArray(serviceList)) {
  throw new Error('Error in JS-Rust-land: serviceList missing or incorrect.');
}

serviceList.some((service) => {
  if (
    typeof service.name !== 'string' ||
    !service.name ||
    (typeof service.url !== 'string' && service.url) ||
    (typeof service.sdl !== 'string' && service.sdl)
  ) {
    throw new Error('Missing required data structure on service.');
  }
});

serviceList = serviceList.map(({ sdl, ...rest }) => ({
  typeDefs: parseTypedefs(sdl),
  ...rest,
}));

function parseTypedefs(source) {
  try {
    return composition.parseGraphqlDocument(source);
  } catch (err) {
    // Return the error in a way that we know how to handle it.
    done({ Err: [err] });
  }
}

try {
  // /**
  //  * @type {{ errors: Error[], supergraphSdl?: undefined, hints: undefined } | { errors?: undefined, supergraphSdl: string, hints: string }}
  //  */
  const composed = composition.composeServices(serviceList);
  done(
    composed.errors
      ? { Err: composed.errors }
      : {
          Ok: {
            supergraphSdl: composed.supergraphSdl,
            hints: composed.hints.map((h) => h.toString()),
          },
        },
  );
} catch (err) {
  done({ Err: [err] });
}
