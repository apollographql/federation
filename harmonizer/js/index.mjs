// This file itself gets its types automatically in VSCode.
//
// However, if you change the exports from this file, be sure to provide the
// types in the do_compose.js which is alongside (and injected in the runtime!)
export { composeServices } from "@apollo/composition";
export { parse as parseGraphqlDocument } from "graphql";
