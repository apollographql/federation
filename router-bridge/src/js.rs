use crate::error::Error;
/// Wraps creating the Deno Js runtime collecting parameters and executing a script.
use deno_core::{op_sync, JsRuntime, RuntimeOptions, Snapshot};
use serde::de::DeserializeOwned;
use serde::Serialize;
use std::sync::mpsc::channel;

pub(crate) struct Js {
    parameters: Vec<(&'static str, String)>,
}

impl Js {
    pub(crate) fn new() -> Js {
        Js {
            parameters: Vec::new(),
        }
    }

    pub(crate) fn with_parameter<T: Serialize>(
        mut self,
        name: &'static str,
        param: T,
    ) -> Result<Js, Error> {
        let serialized = format!(
            "{} = {}",
            name,
            serde_json::to_string(&param).map_err(|error| Error::ParameterSerialization {
                name: name.to_string(),
                message: error.to_string()
            })?
        );
        self.parameters.push((name, serialized));
        Ok(self)
    }

    pub(crate) fn execute<Ok: DeserializeOwned + 'static>(
        &self,
        name: &'static str,
        source: &'static str,
    ) -> Result<Ok, Error> {
        // The snapshot is created in our build.rs script and included in our binary image
        let buffer = include_bytes!("../snapshots/query_runtime.snap");

        // Use our snapshot to provision our new runtime
        let options = RuntimeOptions {
            startup_snapshot: Some(Snapshot::Static(buffer)),
            ..Default::default()
        };
        let mut runtime = JsRuntime::new(options);

        // We'll use this channel to get the results
        let (tx, rx) = channel();

        let happy_tx = tx.clone();

        runtime.register_op(
            "op_result",
            op_sync(move |_state, value, _buffer: ()| {
                happy_tx.send(Ok(value)).expect("channel must be open");

                Ok(serde_json::json!(null))

                // Don't return anything to JS
            }),
        );
        runtime.sync_ops_cache();
        for parameter in self.parameters.iter() {
            runtime
                .execute_script(format!("<{}>", parameter.0).as_str(), &parameter.1)
                .expect("unable to evaluate service list in JavaScript runtime");
        }

        // We are sending the error through the channel already
        let _ = runtime.execute_script(name, source).map_err(|e| {
            let message = format!(
                "unable to invoke {} in JavaScript runtime \n error: \n {:?}",
                source, e
            );

            tx.send(Err(Error::DenoRuntime(message)))
                .expect("channel must be open");

            e
        });

        rx.recv().expect("channel remains open")
    }
}
