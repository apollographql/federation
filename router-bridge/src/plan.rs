/*!
# Create a query plan
*/

use crate::js::Js;
use serde::{de::DeserializeOwned, Deserialize, Serialize};
use std::fmt::{Display, Formatter};
use thiserror::Error;

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
/// Options for the query plan
pub struct QueryPlanOptions {
    /// Use auto fragmentation
    pub auto_fragmentization: bool,
}

/// Default options for query planning
impl QueryPlanOptions {
    /// Default query plan options
    pub fn default() -> QueryPlanOptions {
        QueryPlanOptions {
            auto_fragmentization: false,
        }
    }
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
/// This is the context which provides
/// all the information to plan a query against a schema
pub struct OperationalContext {
    /// The graphQL schema
    pub schema: String,
    /// The graphQL query
    pub query: String,
    /// The operation name
    pub operation_name: String,
}

#[derive(Debug, Error, Serialize, Deserialize, PartialEq)]
/// Container for planning errors
pub struct PlanningErrors {
    /// The contained errors
    pub errors: Vec<PlanningError>,
}

impl Display for PlanningErrors {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        f.write_fmt(format_args!(
            "Planning errors: {}",
            self.errors
                .iter()
                .map(|e| e.to_string())
                .collect::<Vec<String>>()
                .join(", ")
        ))
    }
}

/// An error which occurred during JavaScript planning.
///
/// The shape of this error is meant to mimic that of the error created within
/// JavaScript, which is a [`GraphQLError`] from the [`graphql-js`] library.
///
/// [`graphql-js`]: https://npm.im/graphql
/// [`GraphQLError`]: https://github.com/graphql/graphql-js/blob/3869211/src/error/GraphQLError.js#L18-L75
#[derive(Debug, Error, Serialize, Deserialize, PartialEq)]
pub struct PlanningError {
    /// A human-readable description of the error that prevented planning.
    pub message: Option<String>,
    /// [`PlanningErrorExtensions`]
    pub extensions: Option<PlanningErrorExtensions>,
}

impl Display for PlanningError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        if let Some(msg) = &self.message {
            f.write_fmt(format_args!("{code}: {msg}", code = self.code(), msg = msg))
        } else {
            f.write_str(self.code())
        }
    }
}

#[derive(Debug, Serialize, Deserialize, PartialEq)]
/// Error codes
pub struct PlanningErrorExtensions {
    /// The error code
    pub code: String,
}

/// An error that was received during planning within JavaScript.
impl PlanningError {
    /// Retrieve the error code from an error received during planning.
    pub fn code(&self) -> &str {
        match self.extensions {
            Some(ref ext) => &*ext.code,
            None => "UNKNOWN",
        }
    }
}

/// Create the query plan by calling in to JS.
///
/// We use a generic here because the output type `QueryPlan` is part of the router.
/// Since this bridge is temporary we don't to declare the `QueryPlan` structure in this crate.
/// We will instead let the caller define what structure the plan result should be deserialized into.
pub fn plan<T: DeserializeOwned + 'static>(
    context: OperationalContext,
    options: QueryPlanOptions,
) -> Result<T, PlanningErrors> {
    Js::new()
        .with_parameter("schemaString", context.schema)
        .with_parameter("queryString", context.query)
        .with_parameter("options", options)
        .with_parameter("operationName", context.operation_name)
        .execute("do_plan", include_str!("../js-dist/do_plan.js"))
        .map_err(|errors| PlanningErrors { errors })
}

#[cfg(test)]
mod tests {
    use super::*;

    const SCHEMA: &str = include_str!("testdata/schema.graphql");
    const QUERY: &str = include_str!("testdata/query.graphql");

    #[test]
    fn it_works() {
        insta::assert_snapshot!(serde_json::to_string_pretty(
            &plan::<serde_json::Value>(
                OperationalContext {
                    schema: SCHEMA.to_string(),
                    query: QUERY.to_string(),
                    operation_name: "".to_string()
                },
                QueryPlanOptions::default()
            )
            .unwrap()
        )
        .unwrap());
    }

    #[test]
    fn invalid_schema_is_caught() {
        let result = Err::<String, _>(PlanningErrors {
            errors: vec![PlanningError {
                message: Some("Syntax Error: Unexpected Name \"Garbage\".".to_string()),
                extensions: None,
            }],
        });
        assert_eq!(
            result,
            plan(
                OperationalContext {
                    schema: "Garbage".to_string(),
                    query: QUERY.to_string(),
                    operation_name: "".to_string(),
                },
                QueryPlanOptions::default(),
            )
        );
    }

    #[test]
    fn invalid_query_is_caught() {
        let result = Err::<String, _>(PlanningErrors {
            errors: vec![PlanningError {
                message: Some("Syntax Error: Unexpected Name \"Garbage\".".to_string()),
                extensions: None,
            }],
        });
        assert_eq!(
            result,
            plan(
                OperationalContext {
                    schema: SCHEMA.to_string(),
                    query: "Garbage".to_string(),
                    operation_name: "".to_string(),
                },
                QueryPlanOptions::default(),
            )
        );
    }

    #[test]
    fn query_missing_subfields() {
        let expected_error_message =
            r#"Invalid empty selection set for field "User.reviews" of non-leaf type [Review]"#;
        let result = Err::<String, _>(PlanningErrors {
            errors: vec![PlanningError {
                message: Some(expected_error_message.to_string()),
                extensions: None,
            }],
        });
        // This query contains reviews, which requires subfields
        let query_missing_subfields = "query ExampleQuery { me { id reviews } }".to_string();
        assert_eq!(
            result,
            plan(
                OperationalContext {
                    schema: SCHEMA.to_string(),
                    query: query_missing_subfields,
                    operation_name: "".to_string(),
                },
                QueryPlanOptions::default(),
            )
        );
    }

    #[test]
    fn query_field_that_doesnt_exist() {
        let expected_error_message = r#"Cannot query field "thisDoesntExist" on type "Query"."#;
        let result = Err::<String, _>(PlanningErrors {
            errors: vec![PlanningError {
                message: Some(expected_error_message.to_string()),
                extensions: None,
            }],
        });
        // This query contains reviews, which requires subfields
        let query_missing_subfields = "query ExampleQuery { thisDoesntExist }".to_string();
        assert_eq!(
            result,
            plan(
                OperationalContext {
                    schema: SCHEMA.to_string(),
                    query: query_missing_subfields,
                    operation_name: "".to_string(),
                },
                QueryPlanOptions::default(),
            )
        );
    }
}
