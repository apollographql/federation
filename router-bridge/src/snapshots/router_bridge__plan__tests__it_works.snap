---
source: router-bridge/src/plan.rs
expression: "serde_json::to_string_pretty(&plan::<serde_json::Value>(OperationalContext{schema:\n                                                                               SCHEMA.to_string(),\n                                                                           query:\n                                                                               QUERY.to_string(),\n                                                                           operation:\n                                                                               \"\".to_string(),},\n                                                        QueryPlanOptions::default()).unwrap()).unwrap()"

---
{
  "kind": "QueryPlan",
  "node": {
    "kind": "Fetch",
    "serviceName": "accounts",
    "variableUsages": [],
    "operation": "{me{name{first last}}}"
  }
}
